File: ios/Plugin/OryAuth.swift
import Foundation
import AnyCodable
import SwiftKeychainWrapper


@objc public class OryAuth: NSObject {
    
    let VERIFIED = "VERIFIED" // TODO
    
    @objc public func session(completion: @escaping (_ active: Bool, _ error: Error?) -> Void) -> Void {

        let xSessionToken: String? = SessionTokenStore.get()

        ClientFrontendAPI.toSession(xSessionToken: xSessionToken) { (session, error) in
            guard error == nil else {
                completion(false, error)
                return
            }

            completion(session?.active ?? false, nil)
        }
    }

    @objc public func login(identifier: String, password: String, completion: @escaping (_ error: Error?) -> Void) -> Void {
        
        ClientFrontendAPI.createNativeLoginFlow( returnSessionTokenExchangeCode: true) { (flow, error) in
            guard error == nil else {
                completion(error)
                return
            }
            
            let psswd = UpdateLoginFlowWithPasswordMethod(
                identifier: identifier, method: "password", password: password)
            let body = UpdateLoginFlowBody.typeUpdateLoginFlowWithPasswordMethod(psswd)

            ClientFrontendAPI.updateLoginFlow(flow: flow!.id, updateLoginFlowBody: body) { (login, error) in
                guard error == nil else {
                    completion(error)
                    return
                }
                let saveSuccessful = SessionTokenStore.set(sessionToken: login!.sessionToken!)
                guard saveSuccessful == true else {
                    completion(error) // TODO create new error
                    return
                }
                completion(nil)
            }
        }
        
        return
    }

    @objc public func logout(completion: @escaping (_ error: Error?) -> Void) -> Void {
        let xSessionToken: String? = SessionTokenStore.get()
        let body = PerformNativeLogoutBody.init(sessionToken: xSessionToken!)
        ClientFrontendAPI.performNativeLogout(performNativeLogoutBody: body) { data, error in
            guard error == nil else {
                completion(error)
                return
            }
            
            SessionTokenStore.clear()
            completion(nil)
        }
        return
    }

    @objc public func signin(password: String, traits: String ,completion: @escaping (_ id: String?, _ error: Error?) -> Void) -> Void {
        
        ClientFrontendAPI.createNativeRegistrationFlow(returnSessionTokenExchangeCode: true) { flow, error in
            guard error == nil else {
                completion(nil, error)
                return
            }
            
            let psswd = UpdateRegistrationFlowWithPasswordMethod(
                method: "password", password: password, traits: AnyCodable.init(stringLiteral: traits))
            let body = UpdateRegistrationFlowBody.typeUpdateRegistrationFlowWithPasswordMethod(psswd)
            
            ClientFrontendAPI.updateRegistrationFlow(flow: flow!.id, updateRegistrationFlowBody: body) { registration, error in
                guard error == nil else {
                    completion(nil, error)
                    return
                }
                let saveSuccessful: Bool = SessionTokenStore.set(sessionToken: registration!.sessionToken!)
                guard saveSuccessful == true else {
                    completion(nil, error) // TODO create new error
                    return
                }
                registration?.continueWith?.forEach({ item in
                    switch item {
                    case .typeContinueWithVerificationUi(let value):
                        completion(value.flow.id, nil)
                        return
                    case .typeContinueWithSetOrySessionToken(_):
                        break
                    }
                })
                return completion(nil, nil)
            }
            
        }
        
    }

    @objc public func verify(id: String, code: String, email: String?, completion: @escaping (_ active: Bool, _ error: Error?) -> Void) -> Void {
        
        ClientFrontendAPI.getVerificationFlow(id: id) { flow, error in
            guard error == nil else {
                completion(false, error)
                return
            }
            
            let body = UpdateVerificationFlowBody.typeUpdateVerificationFlowWithCodeMethod(UpdateVerificationFlowWithCodeMethod(code: code, email: email, method: UpdateVerificationFlowWithCodeMethod.Method.link))
            
            ClientFrontendAPI.updateVerificationFlow(flow: flow!.id, updateVerificationFlowBody: body) { verification, error in
                guard error == nil else {
                    completion(false, error)
                    return
                }
                
                completion(verification?.state?.description == self.VERIFIED, nil)
                
            }
        }
    }

}

File: android/src/main/java/es/frnd/capacitor/oryauth/OryAuth.java
package es.frnd.capacitor.oryauth;

import com.getcapacitor.plugin.CapacitorCookieManager;

import java.net.CookieHandler;
import java.util.List;

import sh.ory.ApiClient;
import sh.ory.ApiException;
import sh.ory.api.FrontendApi;
import sh.ory.model.ContinueWith;
import sh.ory.model.ContinueWithVerificationUi;
import sh.ory.model.LoginFlow;
import sh.ory.model.PerformNativeLogoutBody;
import sh.ory.model.RegistrationFlow;
import sh.ory.model.Session;
import sh.ory.model.SuccessfulNativeLogin;
import sh.ory.model.SuccessfulNativeRegistration;
import sh.ory.model.UpdateLoginFlowBody;
import sh.ory.model.UpdateLoginFlowWithCodeMethod;
import sh.ory.model.UpdateLoginFlowWithPasswordMethod;
import sh.ory.model.UpdateRegistrationFlowBody;
import sh.ory.model.UpdateRegistrationFlowWithPasswordMethod;
import sh.ory.model.UpdateVerificationFlowBody;
import sh.ory.model.UpdateVerificationFlowWithCodeMethod;
import sh.ory.model.VerificationFlow;

public class OryAuth {

    private static final String TAG = OryAuth.class.getSimpleName();

    private final SessionTokenStore sessionTokenStore;
    private final FrontendApi frontendApi;

    public OryAuth(SessionTokenStore sessionTokenStore, ApiClient defaultClient) {
        this.sessionTokenStore = sessionTokenStore;
        frontendApi = new FrontendApi(defaultClient);
    }

    public boolean session() {
        String xSessionToken = sessionTokenStore.get();
        if (xSessionToken == null) return false;
        try {
            Session session = frontendApi.toSession(xSessionToken, (String) null, (String) null);

            return Boolean.TRUE.equals(session.getActive());
        } catch (Exception e) {
            return false;
        }
    }

    public String signIn(String password, String traits) throws ApiException {
        RegistrationFlow flow = frontendApi.createNativeRegistrationFlow(true, null);

        UpdateRegistrationFlowWithPasswordMethod psswd = new UpdateRegistrationFlowWithPasswordMethod();
        psswd.setPassword(password);
        psswd.setMethod("password");
        psswd.setTraits(traits);
        UpdateRegistrationFlowBody body = new UpdateRegistrationFlowBody(psswd);

        SuccessfulNativeRegistration registration = frontendApi.updateRegistrationFlow(flow.getId(), body, null);
        String xSessionToken = registration.getSessionToken();
        sessionTokenStore.store(xSessionToken);

        ContinueWithVerificationUi continueWithVerificationUi;
        List<ContinueWith> continueWith = registration.getContinueWith();
        for (int i = 0; i < continueWith.size(); i++) {
            ContinueWith next = continueWith.get(i);
            try {
                continueWithVerificationUi = next.getContinueWithVerificationUi();
                return continueWithVerificationUi.getFlow().getId();
            } catch (ClassCastException ignored) {}
        }
        return null;
    }

    public Object verify(String id, String code, String email) throws ApiException {

        VerificationFlow flow = frontendApi.createNativeVerificationFlow();

        UpdateVerificationFlowWithCodeMethod codeBody = new UpdateVerificationFlowWithCodeMethod().code(code).email(email);
        UpdateVerificationFlowBody body = new UpdateVerificationFlowBody(codeBody);
        VerificationFlow verification = frontendApi.updateVerificationFlow(flow.getId(), body, null, null);

        return verification.getState();
    }

    public boolean logout() throws ApiException {
        String xSessionToken = sessionTokenStore.get();
        if (xSessionToken == null) return false;

        PerformNativeLogoutBody performNativeLogoutBody = new PerformNativeLogoutBody();
        performNativeLogoutBody.setSessionToken(xSessionToken);
        frontendApi.performNativeLogout(performNativeLogoutBody);
        sessionTokenStore.clear();

        return true;
    }

    public void login(String identifier, String password) throws ApiException {

        LoginFlow flow = frontendApi.createNativeLoginFlow(null, null, null, true, null);

        UpdateLoginFlowWithPasswordMethod psswd = new UpdateLoginFlowWithPasswordMethod();
        psswd.setPassword(password);
        psswd.setIdentifier(identifier);
        psswd.setMethod("password");
        UpdateLoginFlowBody body = new UpdateLoginFlowBody(psswd);
        SuccessfulNativeLogin login = frontendApi.updateLoginFlow(flow.getId(), body, null, null);

        String xSessionToken = login.getSessionToken();
        sessionTokenStore.store(xSessionToken);
    }
}
